db/cache_db.go

```diff
@@ -0,0 +1,172 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying database.
+// NOTE: Needed by MultiStore.CacheWrap().
+type CacheDB struct {
+	mtx    sync.Mutex
+	cache  map[string]cDBValue
+	parent DB
+}
+
+// NewCacheDB returns a new in-memory cache.
+func NewCacheDB(parent DB) *CacheDB {
+	db := &CacheDB{
+		cache:  make(map[string]cDBValue),
+		parent: parent,
+	}
+	return db
+}
+
+// Get gets the value for the given key. It retrieves the value from the cache or accesses the
+// underlying database.
+func (db *CacheDB) Get(key []byte) []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	value, ok := db.cache[string(key)]
+	if ok {
+		return value.value
+	}
+	return db.parent.Get(key)
+}
+
+// Set sets the value for the given key in the cache. It does not write the value to the
+// underlying database.
+func (db *CacheDB) Set(key []byte, value []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.SetNoLock(key, value)
+}
+
+// SetSync sets the value for the given key in the cache. It does not write the value to the
+// underlying database.
+func (db *CacheDB) SetSync(key []byte, value []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.SetNoLock(key, value)
+}
+
+// SetNoLock sets the value for the given key in the database. It does not write the value to the
+// underlying database. SetNoLock implements atomicSetDeleter.
```

Comment:
How come we are referring to an `atomicSetDeleter` unexported interface?
Perhaps could we just do a compile type assertion and not mention it? e.g
```go
var _ atomicSetDeleter = (*CacheDB)(nil)
```

Reply:
ok